git_test.go

```diff
@@ -1,35 +1,476 @@
+	if err != nil {
+		fmt.Printf(">> could not initialise test repo")
+		return nil
 	}
 
-	atts := scm.contributeOtelAttributes()
+	return &FakeGitRepo{repo: repo, repoPath: tempDir, t: t}
+}
 
-	assert.Equal(t, 3, len(atts))
+func (r *FakeGitRepo) withBranch(branchName string) *FakeGitRepo {
+	workTree, err := r.repo.Worktree()
+	if err != nil {
+		r.t.Errorf(">> could not get worktree: %v", err)
+		return r
+	}
+
+	err = workTree.Checkout(&git.CheckoutOptions{
+		Branch: plumbing.ReferenceName(branchName),
+		Create: true,
+		Force:  true,
+	})
+	if err != nil {
+		r.t.Errorf(">> could not checkout to branch: %v", err)
+		return r
+	}
+
+	r.divergesFromBase = true
+
+	return r
+}
+
+func (r *FakeGitRepo) addingFile(file string) *FakeGitRepo {
+	workTree, err := r.repo.Worktree()
+	if err != nil {
+		r.t.Errorf(">> could not retrieve worktree: %v", err)
+		return r
+	}
+
+	// copy sample file
+	sampleFile := path.Join(workingDir, file)
+	targetFile := path.Join(tempDir, file)
+
+	sourceFileStat, err := os.Stat(sampleFile)
+	if err != nil {
+		r.t.Errorf(">> could not stat sample file: %v", err)
+		return r
+	}
+
+	if !sourceFileStat.Mode().IsRegular() {
+		r.t.Errorf(">> sample file is not a regular file: %v", err)
+		return r
+	}
+
+	source, err := os.Open(sampleFile)
+	if err != nil {
+		r.t.Errorf(">> could not opem sample file: %v", err)
+		return r
+	}
+	defer source.Close()
+
+	destination, err := os.Create(targetFile)
+	if err != nil {
+		r.t.Errorf(">> could not create target file: %v", err)
+		return r
+	}
+	defer destination.Close()
+
+	_, err = io.Copy(destination, source)
+	if err != nil {
+		r.t.Errorf(">> could not copy file: %v", err)
+		return r
+	}
+
+	_, err = workTree.Add(file)
+	if err != nil {
+		r.t.Errorf(">> could not git-add the file")
+		return r
+	}
+
+	return r
+}
+
+func (r *FakeGitRepo) withCommit(message string) *FakeGitRepo {
+	workTree, err := r.repo.Worktree()
+	if err != nil {
+		r.t.Errorf(">> could not retrieve worktree: %v", err)
+		return r
+	}
+
+	_, err = workTree.Commit(message, &git.CommitOptions{
+		Author: &object.Signature{
+			Name:  "Author Test",
+			Email: "author@test.com",
+			When:  time.Now(),
+		},
+		Committer: &object.Signature{
+			Name:  "Committer Test",
+			Email: "committer@test.com",
+			When:  time.Now(),
+		},
+	})
+	if err != nil {
+		r.t.Errorf(">> could not git-commit the file: %v", err)
+		return nil
+	}
+
+	r.divergesFromBase = true
+
+	return r
+}
+
+func (r *FakeGitRepo) removingFile(file string) *FakeGitRepo {
+	workTree, err := r.repo.Worktree()
+	if err != nil {
+		r.t.Errorf(">> could not retrieve worktree: %v", err)
+		return r
+	}
+
+	targetFile := path.Join(tempDir, file)
+	err = os.Remove(targetFile)
+	if err != nil {
+		r.t.Errorf(">> could not remove target file: %v", err)
+		return r
+	}
+
+	_, err = workTree.Add(file)
+	if err != nil {
+		r.t.Errorf(">> could not git-add the file")
+		return r
+	}
+
+	return r
+}
+
+func (r *FakeGitRepo) read() *GitScm {
```

Comment:
Exit method for the FakeRepo builder, which returns a real SCM object, as those used in the production code